NIFI-5516: Implement Load-Balanced Connections

markap14 opened the pull request:
This is a very involved new feature. At this point, it is functioning well but has the following caveats:
- We need to do a better job of reconciling 'client' with nodes in the cluster for authorization purposes.
- It currently points at a SNAPSHOT version of the nifi-registry. NiFi-Registry will need to be released before this is merged to master. This is due to the VersionedConnection object in the data model being updated such that it now supports configuration of load-balancing properties.

At this point, though, because the changeset is extremely large, I think it makes sense to post a Work-In-Progress PR so that others can begin looking at it. It may also be helpful to reference the Feature Proposal for this feature at https://cwiki.apache.org/confluence/display/NIFI/Load-Balanced+Connections

jtstorck commented:
Reviewing this WIP PR to help get it ready to merge to master when NiFi Registry is released.

pvillard31 commented:
I've been very busy the last few days/weeks and missed that PR. Thanks for the wip @markap14, this is awesome and going to change a lot of things! I gave it a try very quickly and will try to spend more time on it.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-client-dto/src/main/java/org/apache/nifi/web/api/dto/ConnectionDTO.java:
```diff
@@ -231,6 +239,46 @@ public void setPrioritizers(List<String> prioritizers) {
+        this.loadBalanceCompression = compression;
+    }
+
+    @ApiModelProperty(value = "The current status of the Connection's Load Balancing Activities. Status can indicate that Load Balancing is not configured for the connection, that Load Balancing " +
```
Good catch. Will address.

markap14 commented:
@ijokarumawak The scenario that you laid out to drop flowfiles should work just fine. I have tested it several times and just tested again. All FlowFiles were successfully dropped for me. Can you verify that the destination of your connection was stopped and did not have any of the FlowFiles in its possession?
You might also want to try enabling DEBUG logging for org.apache.nifi.controller.queue.SwappablePriorityQueue - it does log quite a few debug statements when performing a drop request.

ijokarumawak commented:
Hi all, I've written some UI code to configure load-balance setting from connection configuration window. Please cherry-pick [this commit](https://github.com/ijokarumawak/nifi/commit/9a3d0b3de8a11120e33ba08400c90a8fbdc93357) to try, it should be easier to test this PR with UI :)
No validation code is added yet. It just sends the configured values as they are.


![image](https://user-images.githubusercontent.com/1107620/45550205-2c586b80-b865-11e8-8cb8-781a1c5e2c41.png)

jtstorck commented:
@markap14 When listing a queue that is load balanced between two nodes in a cluster, I'm seeing duplicate "Position" IDs.  I gather they are unique per node and being represented in the UI as they are reported from each node.  Unique Position IDs should be provided, right?  I don't think we'd necessarily want to add a column for which node the flowfile is on, but that's an additional option.

<img width="1641" alt="screen shot 2018-09-14 at 6 29 53 pm" src="https://user-images.githubusercontent.com/19271493/45577899-5b161800-b84c-11e8-9928-b78e6ac7a8fa.png">

jtstorck commented:
@markap14 I created [PR 5600](https://issues.apache.org/jira/browse/NIFI-5600) to add node information to the flowfile queue display.

jtstorck commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/clustered/partition/RoundRobinPartitioner.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Override
+    public QueuePartition getPartition(final FlowFileRecord flowFile, final QueuePartition[] partitions,  final QueuePartition localPartition) {
+        final long count = counter.getAndIncrement();
```
This counter should probably wrap back to 0 once it reaches max long?

markap14 replied:
It will never reach max long. I assure you that nobody will ever leave an instance of NiFi running long enough to process 2^63 FlowFiles. On very fast system that processes a million FlowFiles per second, that would take 7 million years :)

jtstorck replied:
I think it was Bill Gates that said we'd never need more then 640KB of RAM. :)

joewitt replied:
i am screenshot/saving this thread.  I look forward to coming back to it in less than 7 million years.

markap14 replied:
Ha! I hope to see you in 5 million!

ijokarumawak commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/clustered/client/async/nio/NioAsyncLoadBalanceClientTask.java:
```diff
@@ -0,0 +1,96 @@
+                    final NodeIdentifier clientNodeId = client.getNodeIdentifier();
+                    final NodeConnectionStatus connectionStatus = clusterCoordinator.getConnectionStatus(clientNodeId);
+                    final NodeConnectionState connectionState = connectionStatus.getState();
+                    if (connectionState != NodeConnectionState.CONNECTED) {
```
Do we want to rebalance queued FlowFiles here if strategy allows to do so? FlowFilePartitioner.isRebalanceOnFailure is only evaluated when there is a communication activity. After a node goes down and a remaining node updates the connection status to DISCONNECTED, this branch stops further processing. Thus, FlowFiles will be kept at the connection even though the strategy allows rebalance.

Scenario, with 2 nodes cluster, using RoundRobin, FlowFiles being created continuously and distributed evenly. Then one node goes down. 50% of generated FlowFiles will stay in the connection.

With CorrelationAttributePartitioner, this is the expected behavior. But I expect FlowFiles to be processed at the remaining node in this case with RoundRobin partitioner.

markap14 replied:
That's a good catch.

ijokarumawak commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/clustered/server/ConnectionLoadBalanceServer.java:
```diff
@@ -0,0 +1,251 @@
+                    }
+
+                    logger.debug("Receiving FlowFiles from Peer {}", peerDescription);
+                    loadBalanceProtocol.receiveFlowFiles(socket);
```
Receiving side should implement some way to report that the node cannot receive any more FlowFiles because of back-pressure.

![image](https://user-images.githubusercontent.com/1107620/46066713-14abac00-c1b0-11e8-8b3b-2db86740aa9f.png)


The sending side may keep sending more FlowFiles even the number of queued FlowFiles exceeds the configured back-pressure. Scenario, two-node cluster, nifi0 and nifi1. NiFi0 is the primary, and a connection is configured to use Partition By Attribute. The primary node generates FlowFiles having attribute so that they will be sent to nifi1. NiFi0 only consider the number of queued FlowFile locally to schedule the upstream processor.

markap14 replied:
That's a good point, and honestly not something that I had thought about. Whether or not we would need to test for this would depend on the select partitioner. For example, if trying to offload all data from the node for decommissioning purposes, we want to ignore that, I think, and push data to the node anyway. In the case of Partition By Attribute, though, you have a great point.
Personally, I would be inclined to create another JIRA to address this in the future, and document this as a limitation for the time being. This will take a decent bit of thought to do this efficiently and effectively.

markap14 replied:
In addition to documentation, I do think it would be nice in the UI, when configuring Load Balancing, to indicate that Backpressure Thresholds may not be honored when using load balancing. I also created NIFI-5651 to track this.

markap14 replied:
@ijokarumawak please ignore my comment above :) After thinking about it a good bit more, I realized that the way I was thinking about it previously was far more complicated than it needed to be. I have pushed one more commit that will take backpressure into account.

It works by first checking with the remote node to see if there is space in the queue or if backpressure is applied. If it's applied, the node won't send data. This will cause the data to queue up on the 'client' node, which will cause backpressure to engage there as well.

ijokarumawak commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/clustered/server/StandardLoadBalanceProtocol.java:
```diff
@@ -0,0 +1,578 @@
+                    .setEventDuration(duration)
+                    .setComponentId(connectionId)
+                    .setComponentType("Load Balanced Connection")
+                    .build();
```
![image](https://user-images.githubusercontent.com/1107620/46131222-1770d480-c276-11e8-998c-c4a77853276d.png)

* Although the provenance event is created from a FlowFile, and the FlowFile's content size is set by content claim triple, somehow the provenance shows size as 0.
* `RECEIVE` is reported by `Load Balanced Connection`. I'd expect to the same for the `SEND` event, too. Is this intentional?

markap14 replied:
It is not intentional. Will address.

ijokarumawak commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/clustered/server/StandardLoadBalanceProtocol.java:
```diff
@@ -0,0 +1,578 @@
+                peerDescription, certIdentities);
+
+            authorizer.authorize(certIdentities);
+            logger.debug("Client Identities {} are authorized to load balance data", certIdentities);
```
Just a question, is there any reason for this authorization to be here? I was wondering if it can be done earlier at ConnectionLoadBalanceServer AcceptConnection.run or CommunicateAction.run.

markap14 replied:
I don't think there's any particular reason that I put it here as opposed to the ConnectionLoadBalanceServer...

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/AbstractFlowFileQueue.java:
```diff
@@ -0,0 +1,460 @@
+
+    @Override
+    public synchronized void setLoadBalanceStrategy(final LoadBalanceStrategy strategy, final String partitioningAttribute) {
+        if (strategy == LoadBalanceStrategy.PARTITION_BY_ATTRIBUTE && partitioningAttribute == null) {
```
I'm ok with the ` && !partitioningAttribute.isEmpty()` - but really if partitioning attribute is null or empty and strategy == LoadBalanceStrategy.PARTITION_BY_ATTRIBUTE, we probably should throw an Exception instead.

Re: nullifying this.partitioningAttribute and reseting compression - I disagree. If a user is using Partition By Attribute, then changes it to Round Robin on accident, for instance, and wants to change it back, we should keep the Partitioning Attribute. Or said another way, we should not assume that the user wants the field to be cleared. Similarly, if the user is load balancing, then decides to turn off load balancing (for example to see a difference in performance) we should not reset their configuration. If they want to set it back, they would like (at least I would) expect the previously configured compression value to remain.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/clustered/client/async/nio/RegisteredPartition.java:
```diff
@@ -0,0 +1,69 @@
+        this.flowFileRecordSupplier = flowFileSupplier;
+        this.failureCallback = failureCallback;
+        this.successCallback = successCallback;
+        this.compression = compression;
```
Thanks, I'll check this out. I may have overlooked something.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/clustered/partition/RemoteQueuePartition.java:
```diff
@@ -0,0 +1,337 @@
+                // We've now completed the transaction. We must now update the repositories and "keep the books", acknowledging the FlowFiles
+                // with the queue so that its size remains accurate.
+                updateRepositories(flowFilesSent, Collections.emptyList());
+                priorityQueue.acknowledge(flowFilesSent);
```
I'll review the code here and either address the issue or explain it better.

markap14 replied:
After looking at it more, yes, we should acknowledge in the failure callback as well. Thanks!

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/clustered/partition/FlowFilePartitioner.java:
```diff
@@ -0,0 +1,53 @@
+     * @return <code>true</code> if a change in the size of a cluster should result in re-balancing all FlowFiles in queue,
+     *         <code>false</code> if a change in the size of a cluster does not require re-balancing.
+     */
+    boolean isRebalanceOnClusterResize();
```
I disagree - I think we should stick with standard java convention is `isXYZ()` being a simple method that returns a boolean.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-client-dto/src/main/java/org/apache/nifi/web/api/dto/diagnostics/RemoteQueuePartitionDTO.java:
```diff
@@ -0,0 +1,126 @@
+public class RemoteQueuePartitionDTO {
+    private String nodeId;
+    private int totalFlowFileCount;
+    private long totalByteCount;
```
Wow good catch, I completely overlooked that. Will address.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-client-dto/src/main/java/org/apache/nifi/web/api/dto/diagnostics/LocalQueuePartitionDTO.java:
```diff
@@ -0,0 +1,136 @@
+@XmlType(name = "localQueuePartition")
+public class LocalQueuePartitionDTO {
+    private int totalFlowFileCount;
+    private long totalByteCount;
```
Yup :)

markap14 commented:
@jtstorck @ijokarumawak thanks for the time that you guys have put into reviewing this! I just pushed a new commit, which I believe addresses the feedback above. It also updates the dependency on nifi-registry from 0.3.0-SNAPSHOT to 0.3.0.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FileSystemSwapManager.java:
```diff
@@ -70,9 +71,8 @@
-    private static final Pattern SWAP_FILE_PATTERN = Pattern.compile("\\d+-.+\\.swap");
-    private static final Pattern TEMP_SWAP_FILE_PATTERN = Pattern.compile("\\d+-.+\\.swap\\.part");
+    private static final Pattern SWAP_FILE_PATTERN = Pattern.compile("\\d+-.+(\\..*?)?\\.swap");
+    private static final Pattern TEMP_SWAP_FILE_PATTERN = Pattern.compile("\\d+-.+(\\..*?)?\\.swap\\.part");
```
I don't believe the extra ? is necessary. Either would likely work. I have, however, tested and verified the one that is in the PR. The point of changing it was that the naming convention for swap files changed from "<timestamp>-<uuid>.swap" to "<timestamp>-<uuid>.<partition_name>.swap". The "<partition_name>." is optional in order to maintain backward compatibility in case there are swap files created from a previous version of nifi.

ijokarumawak replied:
The extra ? after '+' is needed if the bracket is meant to capture the optional partition_name part. Anyway, it's not critical since the regex can match with expected strings..

Using regex101.com, we can see the difference:
https://regex101.com/

![image](https://user-images.githubusercontent.com/1107620/46446278-56db7b80-c7b6-11e8-877d-fa653395e68b.png)

![image](https://user-images.githubusercontent.com/1107620/46446268-4cb97d00-c7b6-11e8-86e2-816b68e01fde.png)

markap14 replied:
I get what you're saying now. OK, I agree it makes sense then to add the ?. Will do so. Thanks for pointing that out.

andrewmlim commented on nifi-docs/src/main/asciidoc/administration-guide.adoc:
```diff
@@ -3929,6 +3929,13 @@ from the remote node before considering the communication with the node a failur
+ by the `nifi.cluster.flow.election.max.candidates` property, the cluster will not wait this long. The default value is `5 mins`. Note that the time starts as soon as the first vote is cast.
+|`nifi.cluster.flow.election.max.candidates`|Specifies the number of Nodes required in the cluster to cause early election of Flows. This allows the Nodes in the cluster to avoid having to wait a
+long time before starting processing if we reach at least this number of nodes in the cluster.
+|`nifi.cluster.load.balance.port`|Specifies the port to listen on for incoming connections for load balancing data across the cluster. The default value is `6342`.|
```
Should remove the "|" at the end of line 3936.  Causes incorrect formatting of the property following it.

markap14 replied:
Thanks @andrewmlim! Will do.

ijokarumawak commented:
@markap14 I have done testing this improvements my side. I'm +1 once the StandardFlowSynchronizer mentioned above comment is fixed. Thanks again for this massive improvement, this is going to be really useful!

jtstorck commented:
@markap14 Based on @ijokarumawak's +1, and my testing of this PR along with the changes in my PR for node offloading, I'm a +1 as well.  Will merge this ASAP!

Thanks for this awesome contribution!
